### PR TITLE
[OF-1456] feat: Children now reparent after parent deletion

### PR DESCRIPTION
**Pull request by MAG-mv**
- Updated deletion patch to include children parent changes
- Local entity update callback is called when the children reparent
- Moved resolving of hierarchies into ProcessPendingEntityUpdate to fix issues with self messaging tests
- Added tests

### CODE REVIEW

**Review by MAG-OR — COMMENTED** [review]
LGTM however I don't have detailed knowledge of the space entity code

**Review by MAG-ChristopherAtkinson — DISMISSED** [review]
Changes look good to me, glad to see the edge case is handled for pending entities.

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp [hunk 1]
```diff
@@ -3040,4 +3040,224 @@ CSP_PUBLIC_TEST(CSPEngine, MultiplayerTests, CreateObjectParentTest)
...
+	auto* Connection	 = SystemsManager.GetMultiplayerConnection();
+	auto* EntitySystem	 = SystemsManager.GetSpaceEntitySystem();
+
+	// Log in
+	csp::common::String UserId;
+	LogIn(UserSystem, UserId);
+
+	// Create space
+	const char* TestSpaceName		 = "CSP-UNITTEST-SPACE-MAG";
+	const char* TestSpaceDescription = "CSP-UNITTEST-SPACEDESC-MAG";
```
Comment (MAG-ElliotMorris) [inline Tests/src/PublicAPITests/MultiPlayerTests.cpp:3062]: [Nit] Odd alignment. Could also be `constexpr char*`

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp [hunk 2]
```diff
@@ -3040,4 +3040,224 @@ CSP_PUBLIC_TEST(CSPEngine, MultiplayerTests, CreateObjectParentTest)
...
+
+	char UniqueSpaceName[256];
+	SPRINTF(UniqueSpaceName, "%s-%s", TestSpaceName, GetUniqueString().c_str());
```
Comment (MAG-ElliotMorris) [inline Tests/src/PublicAPITests/MultiPlayerTests.cpp:3065]: [Consider] 
Is C++20 on the table? You might prefer this rather than allocating buffers.
`std::string UniqueSpaceName = std::format("{}-{}", TestSpaceName, GetUniqueString());`

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp [hunk 3]
```diff
@@ -3040,4 +3040,224 @@ CSP_PUBLIC_TEST(CSPEngine, MultiplayerTests, CreateObjectParentTest)
...
+
+	// Test setting the parent for the first child
+	{
+		bool ChildEntityUpdated = false;
+
+		CreatedChildEntity1->SetUpdateCallback(
+			[&ChildEntityUpdated,
+			 ChildEntityName1](SpaceEntity* Entity, SpaceEntityUpdateFlags Flags, csp::common::Array<ComponentUpdateInfo>& UpdateInfo)
+			{
+				if (Entity->GetName() == ChildEntityName1 && Flags & SpaceEntityUpdateFlags::UPDATE_FLAGS_PARENT)
```
Comment (MAG-ElliotMorris) [inline Tests/src/PublicAPITests/MultiPlayerTests.cpp:3101]: [Nit] Stupid tired people like me might thank you for separating logical and bitwise logic more clearly.
```suggestion
				if (Entity->GetName() == ChildEntityName1 && (Flags & SpaceEntityUpdateFlags::UPDATE_FLAGS_PARENT))
```

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp [hunk 4]
```diff
@@ -3040,4 +3040,224 @@ CSP_PUBLIC_TEST(CSPEngine, MultiplayerTests, CreateObjectParentTest)
...
+
+	// Delete the parent
+	{
+		bool DestroyCalled = false;
+
+		auto DestroyCb = [&DestroyCalled](bool Success)
+		{
+			DestroyCalled = true;
+			EXPECT_TRUE(Success);
+		};
```
Comment (MAG-ElliotMorris) [inline Tests/src/PublicAPITests/MultiPlayerTests.cpp:3152]: [Nit] This CB is defined out of line, whilst the update callbacks are all inline? Could be consistent.

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp [hunk 5]
```diff
@@ -3040,4 +3040,224 @@ CSP_PUBLIC_TEST(CSPEngine, MultiplayerTests, CreateObjectParentTest)
...
+		EntitySystem->SetInitialEntitiesRetrievedCallback(EntitiesReadyCallback);
+
+		auto [EnterResult2] = AWAIT_PRE(SpaceSystem, EnterSpace, RequestPredicate, Space.Id);
+		EXPECT_EQ(EnterResult2.GetResultCode(), csp::systems::EResultCode::Success);
+
+		WaitForCallbackWithUpdate(EntitiesCreated, EntitySystem);
+		EXPECT_TRUE(EntitiesCreated);
+	}
+
+	// Ensure children have been parented to the root
```
Comment (MAG-ElliotMorris) [inline Tests/src/PublicAPITests/MultiPlayerTests.cpp:3226]: [Question] Is "parented to the root" the correct terminology, considering that `getParentEntity` is null now? Is this actually a deparent? Maybe :
```suggestion
	// Ensure children have been unparented and are now root entities
```

File: Library/src/Multiplayer/SpaceEntity.cpp [hunk 6]
```diff
@@ -1339,6 +1339,20 @@ void SpaceEntity::ResolveParentChildRelationship()
...
+			// so check pending entities
+			for (const auto& PendingParent : *EntitySystem->PendingAdds)
+			{
+				if (PendingParent->Id == *ParentId)
+				{
+					Parent = PendingParent;
+					break;
+				}
+			}
+		}
```
Comment (MAG-ElliotMorris) [inline Library/src/Multiplayer/SpaceEntity.cpp:1354]: [Consider]
Could `FindSpaceEntityById` be a better place to put this pending check behaviour? It strikes me that this might be the sensible default behaviour?

Perhaps an interface like
`SpaceEntity* SpaceEntitySystem::FindSpaceEntityById(uint64_t EntityId, bool includePending = true);`

**Review by MAG-ElliotMorris — APPROVED** [review]
Just minor nits, nothing blocking. Nice thorough tests.

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp (hunk 2 above)
Comment (MAG-mv) [reply Tests/src/PublicAPITests/MultiPlayerTests.cpp:3065]: I think we're locked to 17 atm

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp (hunk 5 above)
Comment (MAG-mv) [reply Tests/src/PublicAPITests/MultiPlayerTests.cpp:3226]: This is better!

File: Library/src/Multiplayer/SpaceEntity.cpp (hunk 6 above)
Comment (MAG-mv) [reply Library/src/Multiplayer/SpaceEntity.cpp:1354]: I believe this will cause a breaking change for clients as I dont think c++ optionals are supported by the wrapper gen

Comment (MAG-mv) [reply Library/src/Multiplayer/SpaceEntity.cpp:1354]: I want to address the issue with handling entities when we start planning the multiplayer refactor

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp (hunk 4 above)
Comment (MAG-mv) [reply Tests/src/PublicAPITests/MultiPlayerTests.cpp:3368]: This is clang format doing this. It's worth reviewing our usage at some point

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp (hunk 1 above)
Comment (MAG-mv) [reply Tests/src/PublicAPITests/MultiPlayerTests.cpp:3278]: Alignment looks correct on my end. I'm not sure why its being shown like this

Comment (MAG-ChristopherAtkinson) [reply Tests/src/PublicAPITests/MultiPlayerTests.cpp:3278]: I have had issue previously where github is not correctly rendering the file but is fine in checkout, like same thing here

**Review by MAG-OR — COMMENTED** [review]
looks fine to me but I'll let "grown-ups" do the approval

File: Tests/src/PublicAPITests/MultiPlayerTests.cpp (hunk 1 above)
Comment (MAG-ElliotMorris) [reply Tests/src/PublicAPITests/MultiPlayerTests.cpp:3278]: It'll be because github is set to render tabs as 8 spaces by default, your editors probably have it at 4, which aligns correctly.

Imo we should have it so tabs are automatically converted to spaces in our linters, but it's not an issue for right now 🤷

**Review by MAG-AdamThorn — APPROVED** [review]

**Review by MAG-ChristopherAtkinson — APPROVED** [review]
Changed looks good to me since last comment, happy for this to go to approval.
